src/validation/property.rs

```diff
@@ -0,0 +1,266 @@
+        }
+    }
+}
+
+fn validate_name_clashes<T>(validator: &mut Validator, context: &ValidationContext<T>, pou: &PouIndexEntry)
+where
+    T: AnnotationMap,
+{
+    let mut seen = HashSet::new();
+    let mut super_class = pou.get_super_class();
+    while let Some(parent_str) = super_class {
+        if !seen.insert(parent_str) {
+            break;
+        }
+
+        let Some(pou_parent) = context.index.find_pou(parent_str) else {
+            break;
+        };
+
+        // Check if any property in the current POU clashes with a variable in the parent POU
+        for property in pou.get_properties_vec() {
+            if let Some(variable) = context.index.find_local_member(parent_str, &property.ident.name) {
+                validator.push_diagnostic(
+                    Diagnostic::new(format!(
+                        "Name conflict between property and variable `{}` defined in POU `{}` and `{}`",
+                        property.ident.name,
+                        pou_parent.get_name(),
+                        pou.get_name()
+                    ))
+                    .with_error_code("E021")
+                    .with_location(&property.ident.location)
+                    .with_secondary_location(&variable.source_location),
+                );
+            }
+        }
+
+        // Check if any variable in the current POU clashes with a property in the parent POU
+        for member in context.index.get_pou_members(context.qualifier.unwrap_or_default()) {
+            if let Some(property) = pou_parent.get_property(member.get_name()) {
+                validator.push_diagnostic(
+                    Diagnostic::new(format!(
+                        "Name conflict between property and variable `{}` defined in POU `{}` and `{}`",
+                        property.name,
+                        pou_parent.get_name(),
+                        pou.get_name(),
+                    ))
+                    .with_error_code("E021")
+                    .with_location(&property.location)
+                    .with_secondary_location(&member.source_location),
+                );
+            }
+        }
+
+        super_class = pou_parent.get_super_class();
+    }
+}
+
+/// Validates if a derived property is redefined with a conflicting signature
+fn validate_overridden_signatures<T>(
+    validator: &mut Validator,
+    context: &ValidationContext<T>,
+    pou: &PouIndexEntry,
+) where
+    T: AnnotationMap,
+{
+    if pou.get_properties().is_none() {
+        return;
+    }
+
+    let mut seen = HashSet::new();
+    let mut super_class = pou.get_super_class();
+    while let Some(parent_str) = super_class {
+        if !seen.insert(parent_str) {
+            break;
+        }
+
+        let Some(pou_parent) = context.index.find_pou(parent_str) else {
+            break;
+        };
+
+        // No conflicts if one of the two has no properties
+        let Some(properties_child) = pou.get_properties() else {
+            break;
```

Comment:
I think we can't break here. If we have a inhertance chain like A -> B -> C -> D, if B does not have any properties we won't find a conflict between A and D now? Can you double check this? (self-note: maybe a test exists for this and it's actually fine)

Reply:
https://en.meming.world/images/en/f/f7/Well_Yes%2C_But_Actually_No.jpg

On a serious note, `get_properties` currently always returns `Some` if the POU is a Program, FB or Class _even_ if the POU does not contain any properties (i.e. is empty). Anyhow, you're right about the break this should be a continue with the (now) adjusted logic behind `get_properties`